[2단계 - 자동차 경주 리팩터링] 알렉스(이영환) 미션 제출합니다.

yxxnghwan opened the pull request:
소니 1단계 피드백 감사했습니다! 2단계에도 잘 부탁드려요😁
저번 1단계에서 미처 다 고치지 못한 피드백 반영과 MVC 패턴에 맞게 패키지, 클래스, 메서드를 리네임 했습니다.

2단계를 진행하며 들었던 의문점을 몇개 나열해봅니다.

### Input을 담당하는 View 클래스에서 Validator를 호출해도 되는가?

저는 사용자 입력이 검증을 통과하지 못할 경우 에러 메시지를 출력하고 다시 입력받을 수 있게
View의 입력 함수를 재귀로 구현했어요.

재귀 포인트를 잡기 위해 Validator에서 던지는 예외를 캐치해서 메시지 출력하고 재귀시켰는데요,
여기서 View는 비즈니스 로직을 포함하지 않아야 하는데,

- 자동차의 이름이 5글자 이하여야한다.
- 경주 횟수는 자연수여야한다.

등의 조건을 판단하는 부분은 비즈니스 로직이라 생각합니다.
물론 이러한 검증 작업은 Validator가 수행하지만, 
Validator를 View에서 호출하는 것은 View가 비즈니스 로직에 관여하지 않는게 맞나요?

만약 Validator를 호출하는 것도 비즈니스 로직에 관여하는 것이라면,
어느 위치에서 예외를 잡아주는게 적절할까요?

### 이러한 단일 흐름의 어플리케이션에서 Controller의 역할은 무엇인가요?

저는 따로 요청을 처리하는 컨트롤러 클래스를 만들지는 않았습니다...
어떤 역할을 부여해야 하는지를 몰라서 선뜻 만들 수 없었어요
요청에 따라 다른 로직이 실행되는 것이 아니라,
메인 메서드에 주어진 흐름대로 실행되는 단일 흐름의 어플리케이션에서
Controller에겐 어떤 역할을 부여해주어야 하며, 이는 어떤 장점을 가지나요?

당장은 이정도 질문이 있습니다!

이번 리뷰도 잘 부탁드립니다 소니!🤩

yxxnghwan committed:
docs(calculator.README): 구현할 기능 목록 작성

yxxnghwan committed:
test(calculator): StringCalculator 기본 단위 테스트 작성

yxxnghwan committed:
feat(calculator): 0 또는 숫자만 반환하는 경우 작성

yxxnghwan committed:
feat(calculator): 커스텀 구분자가 있는 경우 구분자 추가, 파라미터 추출

yxxnghwan committed:
feat(calculator): 양의 정수가 아닌 경우 예외처리 작성

yxxnghwan committed:
feat(calculator): 숫자 합산

yxxnghwan committed:
fix(calculator): null값이 들어올 경우 NullPointerException 발생하던 버그 수정

yxxnghwan committed:
fix(calculator): 양의 정수 체크 조건을 반대로 적용시킨 버그 수정

yxxnghwan committed:
feat(calculator): main class, UI 구현

yxxnghwan committed:
refactor(calculator): 메소드 분리로 indent 및 라인 수 조절

yxxnghwan committed:
docs(racing-car): 구현할 기능 목록 작성

yxxnghwan committed:
test(racing-car): 난수 생성함수 범위 테스트 작성

yxxnghwan committed:
chore(racing-car): utils 패키지로 이동

yxxnghwan committed:
feat(racing-car): RandomIntegerGenerator 구현

yxxnghwan committed:
docs(racing-car): Exception 종류 수정

yxxnghwan committed:
test(racing-car): CarNameValidatorTest 작성

yxxnghwan committed:
feat(racing-car): CarNameValidator 구현

yxxnghwan committed:
test(racing-car): RoundNumberValidatorTest 작성

yxxnghwan committed:
fix(racing-car): RoundNumberValidatorTest 파라미터 타입 수정

yxxnghwan committed:
feat(racing-car): RoundNumberValidator 구현

yxxnghwan committed:
test(racing-car): CarTest 작성

yxxnghwan committed:
test(racing-car): RacingServiceTest 작성

yxxnghwan committed:
feat(racing-car): Car 초기 구현

yxxnghwan committed:
test(racing-car): CarTest 전진 여부 테스트 추가 검증

랜덤 숫자가 4보다 작을 경우 전진하지 않는지 검사

yxxnghwan committed:
feat(racing-car): RacingService 구현

service 구현에 따른 추가적인 Car.isWinner 메소드 작성

yxxnghwan committed:
feat(racing-car): CarsGenerator 구현

사용자 입력값(String) -> List<Car> 형태로 변환 시키기 위 해 정의

yxxnghwan committed:
test(racing-car): CarsGeneratorTest 작성

yxxnghwan committed:
feat(racing-car): UI 클래스 구현

yxxnghwan committed:
refactor(racing-car): 적절한 파라미터명으로 수정

names -> userInput 변경

yxxnghwan committed:
feat(racing-car): 잘못된 값 입력 시 재입력받도록 수정

yxxnghwan committed:
feat(racing-car): Main 로직 구현

yxxnghwan committed:
refactor(calculator): 리팩토링 및 상수화 작업

yxxnghwan committed:
refactor(calculator): 정규표현식 네이밍 명확하게 수정

yxxnghwan committed:
refactor(racing-car): 리팩토링 및 상수화 작업

yxxnghwan committed:
refactor(calculator): CalculatorInput 클래스 상수화 작업

yxxnghwan committed:
refactor(calculator): StringCalculator 메서드 분리

커스텀 구분자를 가지는 입력인지 판단하는 조건문을 메서드로 분리

yxxnghwan committed:
docs(racing-car): 구현할 기능 목록 수정 (체크박스, 세부적으로 나눔)

구현할 기능 목록에 요소마다 체크 박스를 두고 구현시 체크하도록 변경
한 요소에 여러 기능이 있는 경우 요소를 나눔

yxxnghwan committed:
docs(calculate): 구현할 기능 목록 수정 (체크박스, 세부적으로 나눔)

구현할 기능 목록에 요소마다 체크 박스를 두고 구현시 체크하도록 변경
한 요소에 여러 기능이 있는 경우 요소를 나눔

yxxnghwan committed:
refactor(racing-car): StringCalculator.toNumberList 불필요한 박싱, 언박싱 제거

yxxnghwan committed:
refactor(racing-car): static 키워드 누락되었던 것 추가

yxxnghwan committed:
refactor(racing-car): RacingCarInput 상수선언 개행 변경

yxxnghwan committed:
refactor(racing-car): 옵셔널 처리 변경

yxxnghwan committed:
test(calculator): StringCalculator.calculate 시그니처 변경에 의한 테스트 코드 작성

기존 String 타입을 인자로 받아서 검증, 분리, 계산을 진행
역할을 나눠서 분리된 값을 받아 계산만 진행하도록 클래스 설계를 변경
그에 따른 새로운 테스트 코드 작성

yxxnghwan committed:
test(calculator): StringNumberValidator에 대한 테스트 코드 작성

yxxnghwan committed:
refactor(calculator): StringCalculator에서 검증 로직을 StringNumberValidator로 분리

yxxnghwan committed:
test(calculator): UserInputValidator가 사용자의 입력값이 커스텀 구분자인지 아닌지 판단하는 부분 테스트 코드 작성

yxxnghwan committed:
refactor(calculator): StringCalculator에서 사용자의 입력을 검증하는 기능을 UserInputValidator로 이동

yxxnghwan committed:
test(calculator): UserInputValidator가 null과 whitespace를 잘 걸러내는지 테스트 작성

yxxnghwan committed:
test(calculator): StringCalculator가 수행하던 분리 로직을 수행할 FormulaStringSplitter의 분리 기능 테스트

yxxnghwan committed:
feat(calculator): FormulaStringSplitter 구현

yxxnghwan committed:
feat(calculator): 사용자 입력 형태를 클래스로 정의

yxxnghwan committed:
feat(calculator): 변경된 사항에 따라 CalculatorInput 기능 다시 구현

UserInputValidator의 메서드를 호출하여 검증하고 적절한 UserInputForm을 반환하도록 변경

yxxnghwan committed:
feat(calculator): 변경된 사항에 따라 메인 로직 변경, 콘솔창 입력이 의도대로 되지 않던 오류 수정

yxxnghwan committed:
chore(calculator): 패키지 정리 및 불필요한 import문 삭제

yxxnghwan committed:
style(all): 우테코 스타일 컨벤션 적용

yxxnghwan committed:
refactor(racing-car): 테스트 코드 리팩토링

테스트 메서드 이름을 영어로 붙이고 @DisplayName을 활용해서 warning이 발생하지 않도록 수정

하나의 테스트 메서드에서 여러 케이스를 테스트하는 코드들 각각 테스트 메서드로 분리

yxxnghwan committed:
refactor(racing-car): RacingServiceTest 테스트케이스 준비 과정 분리

@BeforeEach, @AfterEach 활용해서 각 테스트 케이스 시행마다 경주에 참여할 자동차들 세팅

yxxnghwan committed:
refactor(racing-car): RacingService -> RacingGame 으로 변경, 경주 결과를 출력하는 역할을 도메인이 수행하지 않도록 수정

yxxnghwan committed:
refactor(racing-car): Scanner를 RacingCarInput에서만 사용하도록 변경

yxxnghwan committed:
refactor(racing-car): Car 클래스가 randomNumber, winner 등 몰라도 되는 것을 변수명으로 했던 것을 수정

yxxnghwan committed:
chore(racing-car): validator 패키지를 utils 패키지 안으로 이동

yxxnghwan committed:
refactor(racing-car): Car 객체는 랜덤에 대해 아예 모르도록 설계

랜덤은 RacingGame이 뽑아서 Car를 전진시킬지 말지 결정

yxxnghwan committed:
rename(racing-car): 패키지 및 클래스와 메서드 네이밍 수정

yxxnghwan committed:
refactor(racing-car): private 메서드가 public 메서드보다 아래로 가도록 순서 변경

yxxnghwan committed:
refactor(racing-car): 메서드 리팩토링, 상수화 추가

RacingGame.findWinners() forEach에서 스트림의 filter 사용하도록 변경
OutputView에서 우승자 출력부분 delimiter 상수화

yxxnghwan committed:
Merge branch 'yxxnghwan' into step2

sonypark commented on src/main/java/racingcar/domain/Car.java:
```diff
@@ -3,7 +3,6 @@
 public class Car {
-    private static final int PROCEED_FlAG_NUMBER = 4;
 
     private final String name;
```
name이 null로 들어오면 어떻게 되나요?
비즈니스 로직에 대한 검증은 가장 하위 레벨인 도메인 레벨에서 이루어지는게 우선일 것 같아요.
이전 레이어에서 검증을 해주었다고 해도, 언제든 new Car()로 name이 null인 Car가 생성될 여지가 있기 때문입니다.

sonypark commented on src/main/java/racingcar/RacingCarMain.java:
```diff
@@ -1,21 +1,26 @@
-        RacingCarOutput.printRoundResultMessage();
-        racingService.race(round);
-        RacingCarOutput.printWinners(racingService.findWinners());
+        List<Car> cars = InputView.getCars();
```
InputView가 Car를 리턴하는 역할까지 해야할까요?
InputView의 역할이 도메인 객체를 생성하는 역할까지 담당하는게 적절하는게 맞을지도 생각해보면 좋겠네요.
View 레이어는 입력값을 받는 역할을 하고 해당 입력값으로 Car를 생성하는건 도메인 레이어에 넘겨주는건 어떨까요?

sonypark commented on src/main/java/racingcar/domain/RacingGame.java:
```diff
@@ -0,0 +1,42 @@
+        if(RandomIntegerGenerator.random(START_RANDOM_NUMBER, END_RANDOM_NUMBER) > PROCEED_FlAG_NUMBER) {
+            car.proceed();
+        }
+    }
```
1. 자동차가 움직이는 조건이 바뀌면 어떻게 될까요? (ex. 짝수/홀수 인 경우만 전진)
2. race 테스트 코드는 어떻게 작성할 수 있나요?
코드의 변경을 줄이고 테스트를 쉽게 작성할 수 있는 방법은 없을까요?

sonypark commented on src/main/java/racingcar/domain/RacingGame.java:
```diff
@@ -0,0 +1,42 @@
+    public List<Car> race() {
+        cars.forEach(this::randomDrawAndProceed);
+        return cars;
+    }
```
핵심 로직인 것 같은데 테스트 코드가 없네요🥲

sonypark reviewed (changes requested):
알렉스 안녕하세요. 질문 남겨주신 것에 대해 제 생각을 적어봤습니다. 같이 고민해봐요 :)

**1. Input을 담당하는 View 클래스에서 Validator를 호출해도 되는가?**

네 됩니다. 단, 도메인 클래스를 검증하는게 우선입니다. 현재는 View에서만 검증을 하고 있고 domain에서는 검증을 하고 있지 않습니다. 어디서에서 검증이 더 중요할까요? 도메인 객체가 비즈니스 규칙에 맞게 생성됨을 보장하는게 더 중요하지 않을까요? 

알렉스가 View에 검증 로직을 넣어준건 사용자가 잘못 입력했을 때 재귀로 다시 입력하게 해주기 위함일텐데요. 사용자 경험 측면에서 좋은 프로세스라고 생각합니다. 웹에서도 백엔드 단까지 API를 호출하지 않고 프론트에서 유효성 검사로 막는것처럼요.

하지만 View 클래스에서의 검증은 도메인에서의 검증 보다 우선순위가 더 낮다고 생각합니다. View는 언제든 바뀔 수 있고 우회도 가능하기 때문입니다.(console View -> Web View, HTTP API 직접 호출로 검증 로직 우회 등)


**2. 이러한 단일 흐름의 어플리케이션에서 Controller의 역할은 무엇인가요?**

당장의 요구사항에서는 현재 구조는 큰 문제가 없습니다.
현재는 main 메서드에 컨트롤러 로직이 들어가 있는데요. main 메서드에 대한 테스트를 작성하신다면 어떻게 해야할까요? main 메서드에 이러한 로직이 들어가는게 좋은 구조일까요? 컨트롤러는 View에서 입력값을 받아 도메인에 넘겨 비즈니스 로직을 요청하고, 그 응답값을 다시 View에 넘겨주는 역할을 합니다. 테스트 코드를 작성한다고 할 때 main과 controller 중 테스트 코드를 작성하기 용이한 구조는 무엇일까요? 또한, 현재는 컨트롤러가 하나이지만 컨트롤러가 여러개 생긴다면 main 함수의 로직은 더 복잡해지겠죠? MVC가 나온 이유는 무엇인지 한 번 공부해볼까요? +(변경에 유연한 구조, 유지보수에 유리한 구조, 테스트 코드를 작성하기 좋은 구조란 무엇인가?) [참고 - main method vs JUnit](https://techcourse.woowahan.com/s/NASxoOCp/ls/MaJMl3CH)

yxxnghwan committed:
feat(racing-car): 컨트롤러 정의, View 호출 메인 메서드에서 컨트롤러로 이동

yxxnghwan committed:
feat(racing-car): 도메인에서도 검증하도록 변경, View에서 Car 객체를 만들지 않고 Car 이름만 반환하도록 변경

yxxnghwan committed:
refactor(racing-car): RacingGame에서 round를 가지고 Map으로 각 라운드별 결과를 반환하도록 변경

RacingGame.race 메서드를 변경하고 race의 결과를 테스트하는 테스트케이스 작성, 시그니처 변경으로 인한 기존 테스트 케이스 삭제

yxxnghwan committed:
refactor(racing-car): Car의 name과 position을 받는 생성자를 private로 변경하고 deepCopy 메서드 구현, RacingGame의 자동차 전진 메서드 리팩토링

yxxnghwan committed:
style(all): 공백 정리

yxxnghwan committed:
style(all): 공백 정리

yxxnghwan committed:
refactor(racing-car): CarDto 정의, OutputView가 Car를 참조하지 않고 CarDto를 통해서 출력할 수 있도록 변경

yxxnghwan committed:
refactor(racing-car): 메인 메서드에서 직접 도메인을 조작하지 않고 컨트롤러를 거치도록 변경

yxxnghwan committed:
refactor(racing-car): RacingGame.toCarDtos 스트림 사용해서 리팩토링

yxxnghwan committed:
test(racing-car): 공동 우승자 테스트 작성

yxxnghwan committed:
test(racing-car): 컨트롤러 테스트 작성

yxxnghwan commented:
피드백 감사합니다 소니! 덕분에 매번 많이 깨닫고 있어요!🤗
저번 피드백 반영과 새로운 질문사항 뽑아서 PR 보냅니다!
이번엔 MVC 적용과 캡슐화와 테스트 관해서 질문이 꽤 생겼습니다🤔

> 1. Input을 담당하는 View 클래스에서 Validator를 호출해도 되는가?
> 
> 네 됩니다. 단, 도메인 클래스를 검증하는게 우선입니다. 현재는 View에서만 검증을 하고 있고 domain에서는 검증을 하고 있지 않습니다. 어디서에서 검증이 더 중요할까요? 도메인 객체가 비즈니스 규칙에 맞게 생성됨을 보장하는게 더 중요하지 않을까요?
> 
> 알렉스가 View에 검증 로직을 넣어준건 사용자가 잘못 입력했을 때 재귀로 다시 입력하게 해주기 위함일텐데요. 사용자 경험 측면에서 좋은 프로세스라고 생각합니다. 웹에서도 백엔드 단까지 API를 호출하지 않고 프론트에서 유효성 검사로 막는것처럼요.
> 
> 하지만 View 클래스에서의 검증은 도메인에서의 검증 보다 우선순위가 더 낮다고 생각합니다. View는 언제든 바뀔 수 있고 우회도 가능하기 때문입니다.(console View -> Web View, HTTP API 직접 호출로 검증 로직 우회 등)

Car 클래스의 생성자에서 외부에서 들어오는 name과 position을 검증하도록 로직을 추가했습니다.

**질문1** : View에서 1차적으로 하는 검증엔 자동차 이름에 대한 검증 로직이 들어갑니다. 이것은 Car에서 자동차 이름을 검증하는 것과 비슷한 부분이 존재하는데요, CarNameValidator에서 해당 부분만 뽑아 함수로 만들고 재사용 하는게 합리적일까요?

**질문2** : 사용자 입력에 대한 검증은 입력시에 통과했는데, 이후 도메인 검증에서 걸리게 되면 사용자 입력이 잘못된 것이 아니라 시스템의 결함이 됩니다. 이 경우에 예외를 throw 하는 것 말고는 처리할 것이 더 없을까요?

> 2. 이러한 단일 흐름의 어플리케이션에서 Controller의 역할은 무엇인가요?
> 
> 당장의 요구사항에서는 현재 구조는 큰 문제가 없습니다.
> 현재는 main 메서드에 컨트롤러 로직이 들어가 있는데요. main 메서드에 대한 테스트를 작성하신다면 어떻게 해야할까요? main 메서드에 이러한 로직이 들어가는게 좋은 구조일까요? 컨트롤러는 View에서 입력값을 받아 도메인에 넘겨 비즈니스 로직을 요청하고, 그 응답값을 다시 View에 넘겨주는 역할을 합니다. 테스트 코드를 작성한다고 할 때 main과 controller 중 테스트 코드를 작성하기 용이한 구조는 무엇일까요? 또한, 현재는 컨트롤러가 하나이지만 컨트롤러가 여러개 생긴다면 main 함수의 로직은 더 복잡해지겠죠? MVC가 나온 이유는 무엇인지 한 번 공부해볼까요? +(변경에 유연한 구조, 유지보수에 유리한 구조, 테스트 코드를 작성하기 좋은 구조란 무엇인가?)

컨트롤러는 TDD 작성이 비교적 수월하다는 장점이 있군요!
컨트롤러가 모델과 뷰의 다리 역할을 하면, 모델과 뷰를 분리하기 수월해지네요!

> InputView가 Car를 리턴하는 역할까지 해야할까요?
> InputView의 역할이 도메인 객체를 생성하는 역할까지 담당하는게 적절하는게 맞을지도 생각해보면 좋겠네요.
> View 레이어는 입력값을 받는 역할을 하고 해당 입력값으로 Car를 생성하는건 도메인 레이어에 넘겨주는건 어떨까요?

InputView는 입력받은 텍스트를 검증하기 위해 문자열의 컬렉션 까지만 만들어서 검증 후에 반환하도록 변경했습니다.
이후에 RacingGame 에서 생성자로 받아 CarsGenerator를 호출해 cars를 저장하도록 변경했습니다!

> 자동차가 움직이는 조건이 바뀌면 어떻게 될까요? (ex. 짝수/홀수 인 경우만 전진)
> race 테스트 코드는 어떻게 작성할 수 있나요?
> 코드의 변경을 줄이고 테스트를 쉽게 작성할 수 있는 방법은 없을까요?

주신 힌트를 참고해서 메서드 네이밍을 변경했고 가독성과 유지보수를 위해 조건식을 메서드로 분리했습니다.
race의 경우는 각 라운드마다 반복문으로 단건실행이 아니라 RacingGame이 round를 가지고 직접 5라운드를 실행하도록 변경 후에 각 라운드별 경주 결과를 반환하도록 변경하였는데요,
맨 마지막 라운드의 경주 결과를 토대로 maxPosition을 도출하고, findWinners 메서드로 우승자를 도출해서
모든 우승자의 포지션이 maxPosition인지 확인하는 방식으로 테스트 코드를 작성했습니다.

**질문3** : InputView와 OutputView가 두 가지 이상의 역할을 하고 있다고 생각합니다.(예를 들면 InputView는 자동차 이름 입력받기와 라운드 입력받기) View를 나누고 싶은데요, 이 경우에 scanner의 위치는 어디가 좋은지 궁금합니다. 각 InputView가 각자 scanner를 가지나요? 아니면 controller에서 선언하고 View를 호출하는 controller에서 넘겨받는게 좋을까요?

**질문4** : 현재 View 클래스를 정적 메서드를 호출하는 용도로만 사용중인데요, 인스턴스화하여 controller에 컬렉션 형태로 저장하는 방법은 안좋은 방법일까요? 각 View가 한 가지 역할만 하도록 한다면, 한 개의 public 메서드를 가지고 인터페이스를 구현하는 형태로 만들어서 추상화 하고 싶은데요... 그러면 View를 인스턴스화하여 어딘가에 두고 사용해야 하는데 컨트롤러에 컬렉션으로 static final로 선언하는건 안좋은 방법일지 궁금합니다.

**질문5** : 만약 cars를 CarRepository를 만들어서 저장하고 RacingGame을 Service 레이어처럼 사용하면 cars가 오픈되는 샘이 되는데 이렇게 작성하면 테스트 코드를 짜기 더 수월해질거라고 생각이 드는데요(현재 RacingGame에서는 cars에 대한 getter, setter가 없습니다.) 이렇게 구현하는 것도 괜찮은가요? 현재는 메인 메서드에서 주어진 흐름대로 로직이 돌아가지만, 만약 다중 스레드라던가 하는 상황이 있으면 cars에 공통으로 접근해 레이스 컨디션이 유발될 수도 있고, CarRepository의 메서드가 public으로 되어 있을테니 의도치 않은 곳에서 호출할 수 있다는 불안감도 있을 수 있을 것 같아서 선뜻 바꾸지 못했습니다😭

yxxnghwan committed:
test(racing-car): Car 테스트 작성

sonypark commented on src/test/java/racingcar/domain/RacingGameTest.java:
```diff
@@ -0,0 +1,42 @@
+            assertThat(winner.getPosition()).isLessThanOrEqualTo(maxPosition);
+        }
+    }
+
```
`isLessThanOrEqualTo`를 쓰신 이유가 있나요?

sonypark commented on src/main/java/racingcar/domain/RacingGame.java:
```diff
@@ -0,0 +1,62 @@
+        this.cars = CarsGenerator.generateCars(carNames);
+        this.round = round;
+    }
+
+    public Map<Integer, List<CarDto>> race() {
+        Map<Integer, List<CarDto>> raceResults = new LinkedHashMap<>();
+        for (int i = 0; i < round; i++) {
```
👍

sonypark reviewed (changes requested):
알렉스 안녕하세요~
구현 과정에서 여러가지 고민을 하고 시도해보는 것 같아 좋네요💯
미션을 하며 의문을 갖고 스스로에게 질문을 던지고 공부하고 답하는 과정에서 분명 큰 성장이 있을거라 생각합니다. 남겨준 질문에 코멘트 남겼으니 확인부탁드립니다! :)

> 질문1 : View에서 1차적으로 하는 검증엔 자동차 이름에 대한 검증 로직이 들어갑니다. 이것은 Car에서 자동차 이름을 검증하는 것과 비슷한 부분이 존재하는데요, CarNameValidator에서 해당 부분만 뽑아 함수로 만들고 재사용 하는게 합리적일까요?

유틸성 클래스에는 비즈니스 로직과 무관한 기능만 모아놓는게 좋다고 생각합니다. 어느곳에서든 공통적으로 사용할 수 있고 비즈니스 로직이 아닌 기능을 유틸성 클래스에 모아두면 좋을 것 같아요. 현재 뷰에서의 검증은 도메인 검증과 약간 중복이 있더라도 큰 문제 없다고 생각합니다. 중복이 있다고 무조건 분리하는건 오히려 코드 가독성이나 지나친 파편화가 될 수 있다고 생각해요. 이 글도 한 번 읽어보세요~ [Don't make Clean Code harder to maintain, use the Rule of Three](https://understandlegacycode.com/blog/refactoring-rule-of-three/)

> 질문2 : 사용자 입력에 대한 검증은 입력시에 통과했는데, 이후 도메인 검증에서 걸리게 되면 사용자 입력이 잘못된 것이 아니라 시스템의 결함이 됩니다. 이 경우에 예외를 throw 하는 것 말고는 처리할 것이 더 없을까요?

네 도메인 쪽에서 검증에 실패하면 예외를 던져주면 됩니다. 그 예외 메시지는 사용자에게 전달될거고 사용자는 예외를 인지하고 다시 입력하겠죠? 현재는 콘솔기반 어플리케이션이라 도메인에서 예외를 발생했을 때 사용자의 요청을 다시 받게 하기 어려운 부분이 있는데, 웹 어플리케이션을 생각해보면 좀 더 이해하기 쉬울 것 같네요.

> 질문3 : InputView와 OutputView가 두 가지 이상의 역할을 하고 있다고 생각합니다.(예를 들면 InputView는 자동차 이름 입력받기와 라운드 입력받기) View를 나누고 싶은데요, 이 경우에 scanner의 위치는 어디가 좋은지 궁금합니다. 각 InputView가 각자 scanner를 가지나요? 아니면 controller에서 선언하고 View를 호출하는 controller에서 넘겨받는게 좋을까요?

입력을 받는 scanner를 View와 Controller 중 어느 곳에서 갖고 있으면 되는지에 대한 질문 맞을까요? 이건 View의 역할과 Controller의 역할이 무엇인지 생각해보면 쉽게 풀릴 것 같습니다. 어디가 좋을까요?

> 질문4 : 현재 View 클래스를 정적 메서드를 호출하는 용도로만 사용중인데요, 인스턴스화하여 controller에 컬렉션 형태로 저장하는 방법은 안좋은 방법일까요? 각 View가 한 가지 역할만 하도록 한다면, 한 개의 public 메서드를 가지고 인터페이스를 구현하는 형태로 만들어서 추상화 하고 싶은데요... 그러면 View를 인스턴스화하여 어딘가에 두고 사용해야 하는데 컨트롤러에 컬렉션으로 static final로 선언하는건 안좋은 방법일지 궁금합니다.

View를 추상클래스로 만들어서 사용할 때의 장점이 무엇이 있을까요? 모든 View가 공통 메서드를 가져야할 이유가 있는지, 중간에 메서드가 추가될 일은 없을지? 일단 그렇게 한 번 구현해보고 어떤 장단점이 있는지 이야기해주면 좋을 것 같아요~

질문5 : 만약 cars를 CarRepository를 만들어서 저장하고 RacingGame을 Service 레이어처럼 사용하면 cars가 오픈되는 샘이 되는데 이렇게 작성하면 테스트 코드를 짜기 더 수월해질거라고 생각이 드는데요(현재 RacingGame에서는 cars에 대한 getter, setter가 없습니다.) 이렇게 구현하는 것도 괜찮은가요? 현재는 메인 메서드에서 주어진 흐름대로 로직이 돌아가지만, 만약 다중 스레드라던가 하는 상황이 있으면 cars에 공통으로 접근해 레이스 컨디션이 유발될 수도 있고, CarRepository의 메서드가 public으로 되어 있을테니 의도치 않은 곳에서 호출할 수 있다는 불안감도 있을 수 있을 것 같아서 선뜻 바꾸지 못했습니다😭

네 좋은 고민입니다.👍 cars와 관련된 로직을 담당하는 클래스를 만들어 Car의 비즈니스 로직을 담당하게 할 수 있겠네요. 이와 관련해서 [일급컬렉션](https://tecoble.techcourse.co.kr/post/2020-05-08-First-Class-Collection/)이란것에 대해서도 공부해보세요! 단 네이밍이 CarRepository가 적절할지는 고민이 필요해보이네요 :)

yxxnghwan committed:
test(racing-car): 게임 결과 테스트 수정

yxxnghwan committed:
refactor(racing-car): View를 각각 기능에 따라 클래스로 분리

yxxnghwan committed:
refactor(racing-car): 전 커밋에서 CarNamesInputView 빠뜨린거 추가

yxxnghwan committed:
refactor(racing-car): cars를 일급 컬렉션으로 변경

yxxnghwan committed:
refactor(racing-car): 자동차 전진 부분에서 전략 패턴을 활용해서 리팩토링

yxxnghwan committed:
refactor(racing-car): 전진 조건을 RacingGame이 결정하도록 변경

yxxnghwan committed:
test(racing-car): Cars 일급 컬렉션에 대한 테스트 작성

yxxnghwan commented:
안녕하세요 소니, 보내주신 피드백 감사합니다!😁
소니의 피드백 내용과 보내준 문서를 참고해서 고민하고 또 고민했습니다.
과도하게 중복을 제거하려는 것이 유지보수성을 오히려 떨어뜨릴 수도 있겠군요! 현재의 검증이 적당한 것 같습니다.
그리고 일급 컬렉션을 구현해서 저번 질문을 해결했습니다! 일급 컬렉션에 대해 알려주셔서 감사합니다 소니

> isLessThanOrEqualTo를 쓰신 이유가 있나요?
이 부분은 `isEqualTo()`가 더 적절해보이네요! 변경했습니다.

> 입력을 받는 scanner를 View와 Controller 중 어느 곳에서 갖고 있으면 되는지에 대한 질문 맞을까요? 이건 View의 역할과 Controller의 역할이 무엇인지 생각해보면 쉽게 풀릴 것 같습니다. 어디가 좋을까요?
> View를 추상클래스로 만들어서 사용할 때의 장점이 무엇이 있을까요? 모든 View가 공통 메서드를 가져야할 이유가 있는지, 중간에 메서드가 추가될 일은 없을지? 일단 그렇게 한 번 구현해보고 어떤 장단점이 있는지 이야기해주면 좋을 것 같아요~

View를 인터페이스로 묶으면 인스턴스화를 해서 사용해야 하는데 오히려 불편함이 초래됩니다.
View를 나누려고 했을 때 scanner의 위치가 고민되어 InputView라는 인터페이스에 스캐너를 선언하고
상속받은 클래스에서 사용하려고 그러한 구조를 고민했었는데요,
오히려 View를 인스턴스화해서 보관하는 코드를 추가해야하고, 그 공간만큼 메모리도 더 필요하겠더라구요
그리고 그 View를 호출할 Key 타입도 선언했어야 했습니다.🤮

결국 View를 나누긴 했지만, 이전처럼 static 함수만 가지는 형태로 만들었어요.

문제는 CarNamesInputView와 RoundInputView가 각각 스캐너를 선언해서 가지는데,
어플리케이션에 스캐너가 늘어난 샘이 되었습니다. 스캐너의 위치를 계속 고민하고 있어요.🤔

---

자동차의 전진 로직에 대해서 전략 패턴이라는 것을 알게되어 도입해봤습니다.
나중에 전진 조건이 변경되었을 때 더 유연하게 대응할 수 있을 것 같고 테스트를 짜기에도 수월해진 것 같은데
소니의 의견은 어떤가요?

그럼 이번 피드백도 잘 부탁드립니다 소니🤗

yxxnghwan committed:
test(racing-car): ParameterizedTest 사용해서 테스트 코드 줄이기

sonypark commented on src/main/java/racingcar/domain/Cars.java:
```diff
@@ -0,0 +1,61 @@
+                .collect(Collectors.toList());
+    }
+
+    public void proceedAll() {
```
일급컬렉션 사용 👍

sonypark commented on src/main/java/racingcar/domain/ProceedStrategy.java:
```diff
@@ -0,0 +1,5 @@
+package racingcar.domain;
+
+public interface ProceedStrategy {
+    boolean canProceed();
+}
```
전략패턴 💯

sonypark reviewed (approved):
알렉스 안녕하세요!
일급컬렉션과 전략패턴 사용까지..! 훌륭합니다💯
알렉스가 고민한만큼 구조가 개선이 많이 되었네요👍
이번 미션 구현하느라 고생 많으셨습니다!😊
이만 머지하겠습니다. 다음 미션도 화이팅!!💪

yxxnghwan commented:
제 첫 미션 리뷰해주셔서 정말 감사합니다 소니!😁
이리 저리 헤매고 많이 미숙한 저였는데 꼼꼼히 봐주시고 같이 고민해주신 덕분에 미션 마무리할 수 있었던 것 같아요.
앞으로 있을 미션에서도 꾸준히 발전해서 꼭 훌륭한 개발자가 되겠습니다🤙
